kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/analyze/TopicAnalyzeService.java

```diff
@@ -0,0 +1,158 @@
+    analyzeTasksStore.registerNewTask(topicId, task);
+    Schedulers.boundedElastic().schedule(task);
+  }
+
+  public void cancelAnalyze(KafkaCluster cluster, String topicName) {
+    analyzeTasksStore.cancelAnalyze(new TopicIdentity(cluster, topicName));
+  }
+
+  public Optional<TopicAnalyzeStateDTO> getTopicAnalyzeState(KafkaCluster cluster, String topicName) {
+    return analyzeTasksStore.getTopicAnalyzeState(new TopicIdentity(cluster, topicName));
+  }
+
+  public List<TopicAnalyzeStateDTO> getAllTopicAnalyzeStates(KafkaCluster cluster) {
+    return analyzeTasksStore.getAllTopicAnalyzeStates(cluster);
+  }
+
+  class AnalyzingTask implements Runnable, Closeable {
+
+    private final Instant startedAt = Instant.now();
+
+    private final TopicIdentity topicId;
+    private final int partitionsCnt;
+    private final long approxNumberOfMsgs;
+
+    private final TopicAnalyzeStats totalStats = new TopicAnalyzeStats();
+    private final Map<Integer, TopicAnalyzeStats> partitionStats = new HashMap<>();
+
+    private final KafkaConsumer<Bytes, Bytes> consumer;
+
+    AnalyzingTask(KafkaCluster cluster, TopicIdentity topicId, int partitionsCnt, long approxNumberOfMsgs) {
+      this.topicId = topicId;
+      this.approxNumberOfMsgs = approxNumberOfMsgs;
+      this.partitionsCnt = partitionsCnt;
+      this.consumer = consumerGroupService.createConsumer(
+          cluster,
+          // to improve polling throughput
+          Map.of(
+              ConsumerConfig.RECEIVE_BUFFER_CONFIG, "-1", //let OS tune buffer size
+              ConsumerConfig.MAX_POLL_RECORDS_CONFIG, "100000"
+          )
+      );
+    }
+
+    @Override
+    public void close() {
+      consumer.wakeup();
+    }
+
+    @Override
+    public void run() {
+      try {
+        log.info("Starting {} topic analyze", topicId);
+        var topicPartitions = IntStream.range(0, partitionsCnt)
+            .peek(i -> partitionStats.put(i, new TopicAnalyzeStats()))
+            .mapToObj(i -> new TopicPartition(topicId.topicName, i))
+            .collect(Collectors.toList());
+
+        consumer.assign(topicPartitions);
+        consumer.seekToBeginning(topicPartitions);
+
+        var waitingOffsets = new WaitingOffsets(topicId.topicName, consumer, topicPartitions);
+        for (int emptyPolls = 0; !waitingOffsets.endReached() && emptyPolls < 3; ) {
+          var polled = consumer.poll(Duration.ofSeconds(3));
+          emptyPolls = polled.isEmpty() ? emptyPolls + 1 : 0;
+          polled.forEach(r -> {
+            totalStats.apply(r);
+            partitionStats.get(r.partition()).apply(r);
+            waitingOffsets.markPolled(r);
+          });
+          updateProgress();
+        }
+        analyzeTasksStore.setAnalyzeResult(topicId, startedAt, totalStats, partitionStats);
+        log.info("{} topic analyze finished", topicId);
+      } catch (WakeupException | InterruptException cancelException) {
+        log.info("{} topic analyze stopped", topicId);
```

Comment:
do we ever call the `running.remove(topicId);` if we get into this catch clause?

Reply:
it is called in cancel-calling code in `AnalyzeTasksStore::cancelAnalyze `

Reply:
btw, I added `AnalyzeTasksStore:: cancelAnalyze` call here too, just for safety